### PR TITLE
Add tests for device stream

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+   };
+
+test "Local delete device changes appear in v2 /sync",
+   requires => [ local_user_fixtures( 2 ),
+                 qw( can_sync ) ],
+
+   check => sub {
+      my ( $user1, $user2 ) = @_;
+
+      my ( $room_id );
```
Comment (leonerd) [inline tests/41end-to-end-keys/06-device-lists.pl]: Parens aren't necessary here, but don't worry this time.

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+                  or die "user not in changed list";
+
+               Future->done(1);
+            })
+         })
+      });
+   };
+
+
+test "Can query remote device keys using POST after notificaiton",
```
Comment (leonerd) [inline tests/41end-to-end-keys/06-device-lists.pl]: Typo "notification"

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+      });
+   };
+
+
+test "If remote user leaves room we no longer receive device updates",
+   requires => [ local_user_fixture(), remote_user_fixture(), remote_user_fixture(),
+                 qw( can_upload_e2e_keys )],
+
+   check => sub {
+      my ( $user1, $user2, $user3 ) = @_;
```
Comment (leonerd) [inline tests/41end-to-end-keys/06-device-lists.pl]: By the time we get to three users in a test, I usually give them better names. E.g.

    my ( $creator, $joiner, $invitee ) = @_;

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+         do_request_json_for( $user2,
+            method  => "POST",
+            uri     => "/unstable/keys/upload",
+            content => {
+               device_keys => {
+                  user_id => $user2->user_id,
+                  device_id => $user2->device_id,
+               },
+               one_time_keys => {
+                  "my_algorithm:my_id_1", "my+base64+key"
```
Comment (leonerd) [inline tests/41end-to-end-keys/06-device-lists.pl]: Stylistically this wants an `=>` for consistency and readability

**Comment by leonerd** [comment]
Perl syntax and sytest style wise this all looks fine aside one typo in a test title.

Semantics-wise I'm not sure I'm best placed to comment; it's testing a bunch of new API I don't know.

LGTM*

**Comment by erikjohnston** [comment]
Right, I've updated all the styles. I cheekily added one more test I'm afraid, it should have gone on a separate PR but it sort of accidentally ended up in here.

**Pull request merged**